api/workload/v2/stream.go

```diff
@@ -0,0 +1,132 @@
+package workload
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/proto/api/workload"
+	"go.uber.org/multierr"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// streamManager manages connection streams
+type streamManager struct {
+	// Chan is a channel of streams for fetching X509 SVIDs. It is updated whenever a new stream is created.
+	Chan           chan *managedStream
+	ctx            context.Context
+	logger         *logrus.Logger
+	addr           string
+	reconnectChan  chan struct{}
+	connectionChan chan bool
+}
+
+type managedStream struct {
+	workload.SpiffeWorkloadAPI_FetchX509SVIDClient
+
+	closer io.Closer
+}
+
+// Close closes the stream and the underlying connection.
+func (s *managedStream) Close() error {
+	return multierr.Combine(
+		s.SpiffeWorkloadAPI_FetchX509SVIDClient.CloseSend(),
+		s.closer.Close(),
+	)
+}
+
+func newStreamManager(ctx context.Context, logger *logrus.Logger, addr string, connectionChan chan bool) *streamManager {
+	return &streamManager{
+		Chan:           make(chan *managedStream, 1),
+		ctx:            ctx,
+		logger:         logger,
+		addr:           addr,
+		reconnectChan:  make(chan struct{}, 1),
+		connectionChan: connectionChan,
```

Comment:
Moving the "new stream" logic from the reader into the client means that you can drop this channel entirely. As a part of the dealing with the new stream the client can generate a connect event.